sdk/typescript/src/utils/gas-estimation.ts

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { UnserializedSignableTransaction } from '../signers/txn-data-serializers/txn-data-serializer';
+import { Coin, SuiMoveObject } from '../types';
+
+export type TxKind = UnserializedSignableTransaction['kind'];
+type BudgetMap = typeof DEFAULT_GAS_BUDGET_PER_TX_TYPE;
+export type GasBudgetGuessParams<txKind extends TxKind> =
+  txKind extends keyof BudgetMap
+    ? Parameters<BudgetMap[txKind]> extends [boolean, ...infer Params]
+      ? Params
+      : []
+    : [];
+
+export const MIN_GAS_BUDGET = 10;
+export const MAX_GAS_BUDGET = 1_000_000;
+
+const PAY_GAS_FEE_PER_COIN = 150;
+
+/**
+ * Make a guess for the gas budget of a pay tx
+ * @param coins All the coins of the type to send
+ * @param amount The amount to send
+ * @returns The gasBudget guess
+ */
+function computeTransferTxGasBudget(
+  forError: boolean,
+  coins: SuiMoveObject[],
+  amount: bigint
+) {
+  const numInputCoins = Coin.selectCoinSetWithCombinedBalanceGreaterThanOrEqual(
+    coins,
+    amount
+  ).length;
+  let gasBudgetGuess = Math.min(
+    PAY_GAS_FEE_PER_COIN * Math.max(2, Math.min(100, numInputCoins / 2)),
+    MAX_GAS_BUDGET
+  );
+  const isSuiTransfer = Coin.isSUI(coins[0]);
+  if (isSuiTransfer && !forError) {
+    // check if there is enough balance to cover for the amount + the gas
+    // if not lower the gasBudget to allow making better estimations for those cases
+    const totalSuiBalance = coins.reduce(
+      (sum, aCoin) => sum + (Coin.getBalance(aCoin) || BigInt(0)),
+      BigInt(0)
+    );
+    if (totalSuiBalance - BigInt(gasBudgetGuess) - amount < 0) {
+      gasBudgetGuess = Math.max(
+        Number(totalSuiBalance - amount),
+        MIN_GAS_BUDGET
+      );
+    }
+  }
+  return gasBudgetGuess;
+}
+
+const DEFAULT_GAS_BUDGET_PER_TX_TYPE = {
+  mergeCoin: (_forError: boolean) => 10_000,
+  moveCall: (_forError: boolean) => 10_000,
+  pay: computeTransferTxGasBudget,
+  paySui: computeTransferTxGasBudget,
+  payAllSui: computeTransferTxGasBudget,
+  publish: (_forError: boolean) => 10_000,
+  splitCoin: (_forError: boolean) => 10_000,
+  transferObject: (_forError: boolean) => 100,
+  transferSui: (_forError: boolean) => 100,
+} as const;
+
+export function getGasBudgetGuess<T extends TxKind>(
+  txKind: T,
+  maxGasCoinBalance: bigint | number | null,
```

Comment:
why do we want to take maxGasCoinBalance for gas estimation? imo
- estimation is only about txn to be run, if estimation itself failed, we want to fall back to use some constants or numbers in tables as the estimated gas 
- if maxGasCoinBalance < estimated gas (either from dry run or from constants), we want to trigger some preparation steps to make a gas coin for the txn.

Reply:
We are taking `maxGasCoinBalance` into account because that's the minimum budget (not always - in case we send sui) the user can pay. And if we get insufficient funds then we know we need to merge coins. Using as budget only the local constants does't help to improve the estimate.

For example if we set the budget to 1000 and the `maxGasCoinBalance` is  400 the dry run will fail. But the actual cost/budget could be 300. Using `maxGasCoinBalance` helps to improve estimate in this case.

I am not sure if possible, but ideally if we could ignore budget/cost restrictions for dry run would make it easier to use it. To return something like "if you had the gas this would cost x"